tools/slang-cpp-extractor/slang-cpp-extractor-main.cpp

```diff
@@ -0,0 +1,1944 @@
+            {
+                marker = UnownedStringSlice(marker.begin() + m_options.m_prefixMark.getLength(), marker.end() - m_options.m_postfixMark.getLength());
+            }
+            else
+            {
+                marker = UnownedStringSlice::fromLiteral("NORMAL");
+            }
+            out << marker << ", ";
+
+            if (node->m_baseType != Node::BaseType::None || node->m_superNode && node->m_superNode->m_isReflected == false)
+            {
+                out << "BASE, ";
+            }
+            else if (node->hasReflectedDerivedType())
+            {
+                out << "INNER, ";
+            }
+            else
+            {
+                out << "LEAF, ";
+            }
+
+            out << "param) \\\n";
+
+            // Reference the derived types
+            _indent(1, out);
+            out << m_options.m_prefixMark << "DERIVED_" <<  reflectTypeName << "_" << node->m_name.Content << "(x, param) \n";
+            
+            out << "\n";
+        }
+    }
+
+    return SLANG_OK;
+}
+
+SlangResult CPPExtractorApp::calcHeader(CPPExtractor& extractor, StringBuilder& out)
+{
+    const List<Node*>& baseTypes = extractor.getBaseTypes();
+
+    const String& reflectTypeName = m_options.m_reflectType;
+
+    out << "#pragma once\n\n";
+    out << "// Do not edit this file is generated from slang-cpp-extractor tool\n\n";
+
+    for (Index i = 0; i < baseTypes.getCount(); ++i)
+    {
+        Node* baseType = baseTypes[i];
+        if (baseType->m_baseType != Node::BaseType::Marked)
+        {
+            continue;
+        }
+
+        List<Node*> baseScopePath;
+        baseType->calcScopePath(baseScopePath);
+
+        // Remove the global scope
+        baseScopePath.removeAt(0);
+        // Remove the type itself
+        baseScopePath.removeLast();
+
+        for (Node* scopeNode : baseScopePath)
+        {
+            SLANG_ASSERT(scopeNode->m_type == Node::Type::Namespace);
+            out << "namespace " << scopeNode->m_name.Content << " {\n";
+        }
+
+        List<Node*> nodes;
+        baseType->calcDerivedDepthFirst(nodes);
+
+        out << "\n";
+        out << "enum class " << reflectTypeName << "Type\n";
+        out << "{\n";
+
+       
+        Index typeIndex = 0;
+
+        for (Node* node : nodes)
+        {
+            SLANG_ASSERT(node->isClassLike());
+            // If it's not reflected we don't output, in the enum list
+            if (!node->m_isReflected)
+            {
+                continue;
+            }
+
+            // Okay first we are going to output the enum values
+            const Index depth = node->calcDerivedDepth() - 1;
+            _indent(depth, out);
+
+            out << node->m_name.Content << " = " << typeIndex << ",\n";
+
+            typeIndex++;
+        }
+
+        out << "};\n\n";
+
+        out << "\n";
+        out << "enum class " << reflectTypeName << "Last\n";
```

Comment:
This kind of stuff is where the "template engine" style of `slang-generate` could actually make for cleaner code, but as discussed already, entangling this new tool with the existing one would probably complicate things more than necessary for right now.